aeppl/truncation.py

```diff
@@ -0,0 +1,121 @@
+
+    def __init__(self, *args, base_op, **kwargs):
+        self.base_op = base_op
+        super().__init__(*args, **kwargs)
+
+
+@_logprob.register(CensoredRV)
+def censor_logprob(op, value, *inputs, name=None, **kwargs):
+    # TODO: Exact bound check might be problematic
+
+    *rv_params, lower_bound, upper_bound = inputs
+    logprob = _logprob(op.base_op, value, *rv_params, name=name, **kwargs)
+    logcdf = _logcdf(op.base_op, value, *rv_params, name=name, **kwargs)
+
+    is_lower_bounded, is_upper_bounded = False, False
+    if not (
+        isinstance(upper_bound, TensorConstant) and np.all(upper_bound.value == np.inf)
+    ):
+        is_upper_bounded = True
+        logprob = at.switch(
+            at.eq(value, upper_bound),
+            at.log(1 - at.exp(logcdf)),
+            at.switch(at.gt(value, upper_bound), -np.inf, logprob),
+        )
+
+    # Lower censored logp
+    if not (
+        isinstance(lower_bound, TensorConstant) and np.all(lower_bound.value == -np.inf)
+    ):
+        is_lower_bounded = True
+        logprob = at.switch(
+            at.eq(value, lower_bound),
+            logcdf,
+            at.switch(at.lt(value, lower_bound), -np.inf, logprob),
+        )
+
+    if is_lower_bounded and is_upper_bounded:
+        logprob = Assert("lower_bound <= upper_bound")(
+            logprob, at.all(at.le(lower_bound, upper_bound))
+        )
+
+    if name:
+        logprob.name = f"{name}_censored_logprob"
+
+    return logprob
+
+
+@local_optimizer(tracks=[Elemwise])
+def censor_rvs(fgraph: FunctionGraph, node: Node) -> Optional[List[Node]]:
+    # TODO: Canonicalize x[x>ub] = ub -> clip(x, x, ub)
+
+    rv_map_feature = getattr(fgraph, "preserve_rv_mappings", None)
+
+    if rv_map_feature is None:
+        return
+
+    if not isinstance(node.op, Elemwise) or not isinstance(node.op.scalar_op, Clip):
+        return
+
+    clipped_var = node.out
+    if clipped_var not in rv_map_feature.rv_values:
+        return
+
+    base_var, lower_bound, upper_bound = node.inputs
+    base_op = base_var.owner.op
+
+    if not isinstance(base_op, RandomVariable):
+        return
+
+    # base_var var already has a direct value var
+    # TODO: Raise informative Error
+    if base_var in rv_map_feature.rv_values:
+        return
+
+    is_lower_bounded, is_upper_bounded = False, False
+    if lower_bound is not base_var:
+        # y = clip(x, x, ?)
+        is_lower_bounded = True
+
+    if upper_bound is not base_var:
+        # y = clip(x, ?, x)
+        is_upper_bounded = True
+
+    censored_rv = CensoredRV(
+        "censored",
+        base_op.ndim_supp,
+        list(base_op.ndims_params) + [0, 0],
+        base_op.dtype,
+        inplace=False,
+        base_op=base_op,
+    )
+
+    censored_node = censored_rv.make_node(
+        *base_var.owner.inputs,
+        lower_bound if is_lower_bounded else -np.inf,
+        upper_bound if is_upper_bounded else np.inf,
```

Comment:
These lines look problematic: `-np.inf` should have the same shape as `[lower|upper]_bound`